### PR TITLE
enh(api) : add the posibility to search using host groups

### PR DESCRIPTION
**Pull request by Exodiaaaa**
## Description

Enhance endpoint GET: /configuration/hosts/categories by adding possibility to search by host groups

**Fixes** # (MON-34389)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [X] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [X] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>

try to filter hostgroups using the new option on the filter : http://localhost:8080/centreon/api/v24.04/configuration/hosts/categories?search={"hostgroups_id":1}
You should create a hostgroups related to hostcategories to test it

#### Community contributors & Centreon team

- [X] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [X] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 1]
```diff
@@ -104,10 +104,25 @@ public function findAll(?RequestParametersInterface $requestParameters): array
         $concatenator = new SqlConcatenator();
         $concatenator->withCalcFoundRows(true);
         $concatenator->defineSelect(
-            'SELECT hc.hc_id, hc.hc_name, hc.hc_alias, hc.hc_activate, hc.hc_comment
+            'SELECT DISTINCT hc.hc_id, hc.hc_name, hc.hc_alias, hc.hc_activate, hc.hc_comment
             FROM `:db`.hostcategories hc'
         );
 
+        if (! empty($_GET['search']) && mb_strpos($_GET['search'], 'host_group_id')) {
```
Comment (jeremyjaouen) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:111]: Do not use $_GET directly, the REquestParametersInterface has a method to do this check

https://github.com/centreon/centreon/blob/develop/centreon/src/Centreon/Domain/RequestParameters/Interfaces/RequestParametersInterface.php#L52

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

**Review by tuntoja — APPROVED** [review]
Approved GHA related changes.

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 2]
```diff
@@ -508,7 +508,22 @@ private function retrieveHostCategories(
...
+            $concatenator->appendJoins(
+                <<<'SQL'
+                        LEFT JOIN  hostcategories_relation hcr
+                            ON hc.hc_id = hcr.hostcategories_hc_id
+                        LEFT JOIN host h
+                            ON hcr.host_host_id = h.host_id
+                        LEFT JOIN hostgroup_relation hgr
+                            ON h.host_id = hgr.host_host_id
+                        LEFT JOIN hostgroup hg
+                            ON hgr.hostgroup_hg_id = hg.hg_id
```
Comment (a-launois) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:523]: ⚠️ If you are in an on-prem platform with non-admin user, this request is too open, 
you have to filter hostgroups, host and hostcategories on the user's ACLs

I talked about it with @dmyios, you can see with him for more detailed explanations

**Review by a-launois — CHANGES_REQUESTED** [review]

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 3]
```diff
@@ -583,4 +619,41 @@ private function hasRestrictedAccessToHostCategories(array $accessGroupIds): boo
...
+        $concatenator->storeBindValueMultiple(':access_group_ids', $accessGroupIds, \PDO::PARAM_INT)
+            ->appendWhere('ag.acl_group_id IN (:access_group_ids)');
+
+        $statement = $this->db->prepare($this->translateDbName($concatenator->__toString()));
+
+        $concatenator->bindValuesToStatement($statement);
+        $statement->execute();
+
+        return (bool) $statement->fetchColumn();
+    }
```
Comment (a-launois) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:663]: Host group ACLs doesn't work on a "filter" system like the categories, but on the UI you can check a checkbox that gives you access to all host group. You need to check if this option with the hasAccessToAllHostGroups method

```suggestion
```

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 4]
```diff
@@ -150,6 +171,24 @@ public function findAllByAccessGroups(array $accessGroups, ?RequestParametersInt
...
+                        INNER JOIN `:db`.acl_res_group_relations argr
+                            ON res.acl_res_id = argr.acl_res_id
+                        INNER JOIN `:db`.acl_groups ag
+                            ON argr.acl_group_id = ag.acl_group_id
+                        WHERE ag.acl_group_id IN (:access_group_ids)
+                    )
+                   SQL
+            );
+        }
+
```
Comment (a-launois) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:191]:
```suggestion
            INNER JOIN `:db`.acl_groups ag
                ON argr.acl_group_id = ag.acl_group_id'
            LEFT JOIN `:db`.hostcategories_relation hcr
                ON hcr.host_host_id = host.host_id
            LEFT JOIN `:db`.host
                ON host.host_id = hsr.host_host_id
                AND host.host_register = '1'
            LEFT JOIN `:db`.hostgroup_relation hgr
                ON hgr.host_host_id = host.host_id
                {$hostGroupAcls}
            LEFT JOIN `:db`.hostgroup
                ON hostgroup.hg_id = hgr.hostgroup_hg_id
            SQL
        );
```

**Review by a-launois — CHANGES_REQUESTED** [review]

**Review by kduret — DISMISSED** [review]
ok for CI changes

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 5]
```diff
@@ -104,10 +104,25 @@ public function findAll(?RequestParametersInterface $requestParameters): array
         $concatenator = new SqlConcatenator();
         $concatenator->withCalcFoundRows(true);
         $concatenator->defineSelect(
-            'SELECT hc.hc_id, hc.hc_name, hc.hc_alias, hc.hc_activate, hc.hc_comment
+            'SELECT DISTINCT hc.hc_id, hc.hc_name, hc.hc_alias, hc.hc_activate, hc.hc_comment
```
Comment (dmyios) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:107]: No need for `DISTINCT` here

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 6]
```diff
@@ -104,10 +104,25 @@ public function findAll(?RequestParametersInterface $requestParameters): array
...
             FROM `:db`.hostcategories hc'
         );
 
+        if (mb_strpos($requestParameters?->getSearchAsString() ?? '' , 'host.group.id')) {
```
Comment (dmyios) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:111]: Innecessary `if` statement, just append joins to you concatenator

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 7]
```diff
@@ -136,20 +151,50 @@ public function findAllByAccessGroups(array $accessGroups, ?RequestParametersInt
...
-            FROM `:db`.hostcategories hc
-            INNER JOIN `:db`.acl_resources_hc_relations arhr
-                ON hc.hc_id = arhr.hc_id
-            INNER JOIN `:db`.acl_resources res
-                ON arhr.acl_res_id = res.acl_res_id
-            INNER JOIN `:db`.acl_res_group_relations argr
-                ON res.acl_res_id = argr.acl_res_id
-            INNER JOIN `:db`.acl_groups ag
-                ON argr.acl_group_id = ag.acl_group_id'
+            <<< 'SQL'
```
Comment (dmyios) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:174]:
```suggestion
            <<< SQL
```

**Review by dmyios — CHANGES_REQUESTED** [review]

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 8]
```diff
@@ -104,10 +104,25 @@ public function findAll(?RequestParametersInterface $requestParameters): array
...
+            $concatenator->appendJoins(
+                <<<'SQL'
+                        INNER JOIN  hostcategories_relation hcr
+                            ON hc.hc_id = hcr.hostcategories_hc_id
+                        INNER JOIN host h
+                            ON hcr.host_host_id = h.host_id
+                        INNER JOIN hostgroup_relation hgr
+                            ON h.host_id = hgr.host_host_id
+                        INNER JOIN hostgroup hg
+                            ON hgr.hostgroup_hg_id = hg.hg_id
```
Comment (dmyios) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:121]: Use rather `LEFT JOIN`s

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 9]
```diff
@@ -136,20 +151,50 @@ public function findAllByAccessGroups(array $accessGroups, ?RequestParametersInt
...
+                    INNER JOIN `:db`.hostcategories_relation hcr
+                        ON hcr.host_host_id = host.host_id
+                    INNER JOIN `:db`.host
+                        ON host.host_id = hsr.host_host_id
+                        AND host.host_register = '1'
+                    INNER JOIN `:db`.hostgroup_relation hgr
+                        ON hgr.host_host_id = host.host_id
+                        {$hostGroupAcls}
+                    INNER JOIN `:db`.hostgroup
+                        ON hostgroup.hg_id = hgr.hostgroup_hg_id
```
Comment (dmyios) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:194]: Use rather `LEFT JOIN`s

**Review by dmyios — CHANGES_REQUESTED** [review]

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 10]
```diff
@@ -107,6 +107,18 @@ public function findAll(?RequestParametersInterface $requestParameters): array
             'SELECT hc.hc_id, hc.hc_name, hc.hc_alias, hc.hc_activate, hc.hc_comment
             FROM `:db`.hostcategories hc'
         );
+        $concatenator->appendJoins(
```
Comment (jeremyjaouen) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:110]: Like the other PR, check that the filter is present before doing the join

**Review by a-launois — COMMENTED** [review]
I suggest that you check what Dmytro has done on this two PRs:
https://github.com/centreon/centreon/pull/3351
https://github.com/centreon/centreon/pull/3407
The main attention points being the ACLs 
- on host (check if user has access to all hosts check) 
- as well as groups/categories (check if user has filtered access)

**Review by a-launois — CHANGES_REQUESTED** [review]
I suggest that you check what Dmytro did on this two PRs:
https://github.com/centreon/centreon/pull/3351
https://github.com/centreon/centreon/pull/3407
The main attention points being the ACLs

on host (check if user has access to all hosts check)
as well as groups/categories (check if user has filtered access)

**Comment by adr-mo** [comment]
Closing this PR as I take back the subject. Thanks @Exodiaaaa for the help 💯

**Pull request closed without merging**